### PR TITLE
[MNG-7390] Allow selecting modules outside the cwd into the reactor using --projects.

### PR DESCRIPTION
**Pull request by MartinKanters**
There is one main functional change in this PR which is described in the following ticket: https://issues.apache.org/jira/browse/MNG-7390.

There are some refactorings and cleanups done as well:
- It used to sort the intermediate project list a couple of times which were not needed
- It used to calculate and add the also-make/also-make-dependent projects into the intermediate project list.
- The main "reactor" method (`DefaultGraphBuilder#reactorDependencyGraph`) is made more descriptive.. that one method should describe exactly every business logic rule, instead of having business rules inside methods. In total I think it's easier to understand to full flow now.
- The methods used for getting and trimming the intermediate project list were made consistent. They now additionally:
  - fail fast when it's not applicable,
  - return Sets, as the project lists do not need to be sorted in between,
  - are converted a bit more to Streams for the sake of readability.

The relevant IT is found here: https://github.com/apache/maven-integration-testing/pull/135
Jira issue: https://issues.apache.org/jira/browse/MNG-7390

---

Following this checklist to help us incorporate your
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MNG) filed
       for the change (usually before you start working on it).  Trivial changes like typos do not
       require a JIRA issue. Your pull request should address just this issue, without
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MNG-XXX] SUMMARY`, where you replace `MNG-XXX`
       and `SUMMARY` with the appropriate JIRA issue. Best practice is to use the JIRA issue
       title in the pull request title and in the first line of the commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will
       be performed on your pull request automatically.
 - [x] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

**Review by michael-o — CHANGES_REQUESTED** [review]
Please select a proper commit title because it is now disjoint with the JIRA summary. Which one is better from your PoV?

**Comment by MartinKanters** [comment]
> Please select a proper commit title because it is now disjoint with the JIRA summary. Which one is better from your PoV?

Good point! After some iterations this is what I came up with. The previous one contained something like "usual reactor", which is super vague. What do you think?

**Comment by MartinKanters** [comment]
> The question is: Did you fix a cwd traversal or are you able to discover the root reactor although you are in a module?

The latter, it discovers the reactor as if you were running it from the root, that's why you can select modules from different directories inside the same project.

**Comment by michael-o** [comment]
> > The question is: Did you fix a cwd traversal or are you able to discover the root reactor although you are in a module?
> 
> The latter, it discovers the reactor as if you were running it from the root, that's why you can select modules from different directories inside the same project.

under every condition or does it strictly require a `.mvn/` to be present?

File: maven-core/src/main/java/org/apache/maven/graph/DefaultGraphBuilder.java [hunk 1]
```diff
@@ -242,106 +260,86 @@ public DefaultGraphBuilder( BuildResumptionDataRepository buildResumptionDataRep
...
-
-            // Order the new list in the original order
-            List<MavenProject> sortedProjects = graph.getSortedProjects();
-            result.sort( comparing( sortedProjects::indexOf ) );
+            return Collections.emptySet();
         }
 
-        return result;
+        List<MavenProject> emptyList = Collections.emptyList();
+        return projectsToActivate.stream()
```
Comment (mthmulders) [inline maven-core/src/main/java/org/apache/maven/graph/DefaultGraphBuilder.java:337]: I _think_ this does the same as the previous code, but to be honest, I'm having a hard time to make sure it does. Why did you introduce a bunch of Streams here? Does it give any benefit over the previous, Collection-based approach?

**Comment by MartinKanters** [comment]
> I see a lot of code has changed due to a conversion from `List` to `Set` (which I think is fine, but I'm not 100% sure) and sometimes a variable rename. Unfortunately, it's all one big commit, which makes it hard for me to focus on the essentials of the change.

I understand, it became quite terrible to review, sorry for that. The change in essence is very simple. 
This is about it: https://github.com/apache/maven/compare/master...MartinKanters:MNG-7390-select-module-outside-pwd-simplified

> Is the `List` to `Set` conversion _necessary_ for this feature to work? For the variable rename, I can answer the question myself... ;-)

I took the opportunity to refactor the whole class, as I never really liked the coding style much. It's hard to grasp the full business logic and since we have covered it with a lot of unit tests and integration tests I had confidence that I could make the refactor. I've put a list of main things changed in the PR description. If you want, we can have a meeting to go over the points together, let me know. :)

File: maven-core/src/main/java/org/apache/maven/graph/DefaultGraphBuilder.java (hunk 1 above)
Comment (MartinKanters) [reply maven-core/src/main/java/org/apache/maven/graph/DefaultGraphBuilder.java:337]: Mostly readability, but I'm okay to revert that if you disagree that it reads better.

File: maven-core/src/main/java/org/apache/maven/graph/DefaultGraphBuilder.java [hunk 2]
```diff
@@ -137,72 +139,88 @@ public DefaultGraphBuilder( BuildResumptionDataRepository buildResumptionDataRep
...
-                result = includeAlsoMakeTransitively( result, request, graph );
+        Set<MavenProject> selectedProjects = new HashSet<>( requiredSelectors.size() + optionalSelectors.size() );
+        selectedProjects.addAll( getProjectsBySelectors( request, allSortedProjects, requiredSelectors, true ) );
+        selectedProjects.addAll( getProjectsBySelectors( request, allSortedProjects, optionalSelectors, false ) );
 
-                // Order the new list in the original order
-                List<MavenProject> sortedProjects = graph.getSortedProjects();
-                result.sort( comparing( sortedProjects::indexOf ) );
-            }
+        // it can be empty when an optional project is missing from the reactor
```
[the remark this answers is not in the document]
Comment (MartinKanters) [reply maven-core/src/main/java/org/apache/maven/graph/DefaultGraphBuilder.java:213]: Actually this if-check is redundant. Basically it does: if (set == empty) return emptySet(); else return set.
I'll remove it :)

**Comment by michael-o** [comment]
The changeset is quite large, I am having a hard to to review it properly. I will leave the review to @mthmulders for now and then will check again when you both agree.

**Comment by mthmulders** [comment]
> The change in essence is very simple.
This is about it: https://github.com/apache/maven/compare/master...MartinKanters:MNG-7390-select-module-outside-pwd-simplified

Thanks, that is very helpful! The essence of the change is indeed pretty simple and I can make sense out of that.

> [...] since we have covered it with a lot of unit tests and integration tests I had confidence that I could make the refactor

That's what we have those tests for :-)

I'm OK with the changes in the "simplified" version. Given that the tests all pass, I'm fine with merging the changes.

Just a thought: would it be feasible to merge the simplified changes, and then have the refactoring as a separate PR, just to make it easier for everybody to follow along and review? I'd prefer to keep everyone involved as this is a piece of the code which is pretty hard to understand in itself...

**Comment by MartinKanters** [comment]
> > The change in essence is very simple.
> > This is about it: [master...MartinKanters:MNG-7390-select-module-outside-pwd-simplified](https://github.com/apache/maven/compare/master...MartinKanters:MNG-7390-select-module-outside-pwd-simplified)
> 
> Thanks, that is very helpful! The essence of the change is indeed pretty simple and I can make sense out of that.
> 

> > [...] since we have covered it with a lot of unit tests and integration tests I had confidence that I could make the refactor
> 
> That's what we have those tests for :-)
> 
> I'm OK with the changes in the "simplified" version. Given that the tests all pass, I'm fine with merging the changes.
> 
> Just a thought: would it be feasible to merge the simplified changes, and then have the refactoring as a separate PR, just to make it easier for everybody to follow along and review? I'd prefer to keep everyone involved as this is a piece of the code which is pretty hard to understand in itself...

Sure, that's fine, I can do that. I'll try to split the changes into separate commits for review sake.
It will take some time, though, I don't have that much time nowadays.. :) 

For now I'll create a separate PR for the simplified version (and ensure that the IT branch is renamed to the simplified branch name as well. Then I'll close this PR.

**Comment by michael-o** [comment]
Dank U

**Comment by MartinKanters** [comment]
> Dank U

Bitte schön.

---

Here's the simplified PR, currently running against the ITs: https://github.com/apache/maven/pull/677
Related PR for maven-integration-tests (branch has been renamed to match this PR's branch): https://github.com/apache/maven-integration-testing/pull/138

Will now close this PR and later create a new one just for the refactorings.

**Comment by jira-importer** [comment]
Resolve #9175

**Pull request closed without merging**